### PR TITLE
Network: Implement TCP based communication

### PR DESCRIPTION
**Pull request by omerfirmak**
Part of #203

### CODE REVIEW

**Comment by omerfirmak** [comment]
I'll enabled this to be used on integration tests, mixed with HTTP interfaces.

File: tests/system/node/0/config.yaml [hunk 1]
```diff
@@ -59,8 +59,10 @@ validator:
 ################################################################################
 network:
   # Supported value: IPv4, IPv6
-  - https://node-2:2735
-  - https://node-3:3735
-  - https://node-4:4735
-  - https://node-5:5735
-  - https://node-6:6735
+  - tcp://node-0:1735
```
Comment (Geod24) [inline tests/system/node/0/config.yaml]: Self connection ?

File: source/agora/network/RPC.d [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+                log.trace("[SERVER] Done writing...");
+            }
+            return;
+        }
+    default:
+        assert(0);
+    }
+}
+
+// /// Generate an enum used for the dispatch of functions
```
Comment (kchulj) [inline source/agora/network/RPC.d]: Are these implementations for the future?

File: source/agora/network/RPC.d [hunk 3]
```diff
@@ -0,0 +1,498 @@
+/*******************************************************************************
+
+    An RPC generator, similar to `vibe.web.rest`, but for efficient node-to-node
+    communication.
+
+    The aim of this module is to provide a low-overhead way to interface nodes.
+    This module uses TCP for communication and binary data serialization.
+    It should also be fairly simple to chain generator, e.g. writting a server
```
Comment (kchulj) [inline source/agora/network/RPC.d]:
```suggestion
    It should also be fairly simple to chain generator, e.g. writing a server
```

**Comment by kchulj** [comment]
Is there a reason that the 4th commit is separate?

**Comment by Geod24** [comment]
> tests/system/source/main.d(54,24): Error: template instance `RestInterfaceClient!API` template `RestInterfaceClient` is not defined

Integration test client does not compile.

File: source/agora/network/RPC.d [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+    An RPC generator, similar to `vibe.web.rest`, but for efficient node-to-node
+    communication.
+
+    The aim of this module is to provide a low-overhead way to interface nodes.
+    This module uses TCP for communication and binary data serialization.
+    It should also be fairly simple to chain generator, e.g. writing a server
+    that receives `vibe.web.rest` queries and forwards them is trivial.
+
+    Copyright:
+        Copyright (c) 2019-2020 BOS Platform Foundation Korea
```
Comment (Geod24) [inline source/agora/network/RPC.d]:
```suggestion
        Copyright (c) 2019-2021 BOS Platform Foundation Korea
```

File: source/agora/network/RPC.d [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+    }
+
+    ubyte[1024] buffer = void;
+    scope DeserializeDg reader = (size_t size) @safe
+    {
+        assert(size < buffer.length);
+        scope (failure) assert(0);
+        stream.read(buffer[0 .. size]);
+        return buffer[0 .. size];
+    };
```
Comment (Geod24) [inline source/agora/network/RPC.d]: Hum I wonder if 1024 will not cause problem. Looks like an obvious DoS vector.

File: source/agora/network/RPC.d [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+                        conn.read(method[0 .. $]);
+                        this.log.trace("[CLIENT] Returning from {} : {}",
+                                       __PRETTY_FUNCTION__, method);
+                    }
+                    else
+                    {
+                        scope DeserializeDg dg = (size_t size)
+                            {
+                                // this.log.info("[{}] Deserializing {}",
+                                //     Ret.stringof, size);
```
Comment (Geod24) [inline source/agora/network/RPC.d]: Remove ?

File: source/agora/network/RPC.d [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+                                if (size >= tmp.length)
+                                    throw new Exception("Out of bound read");
```
Comment (Geod24) [inline source/agora/network/RPC.d]: `ensure`

File: source/agora/network/RPC.d (hunk 2 above)
Comment (Geod24) [reply source/agora/network/RPC.d]: Those should be removed.

**Review by Geod24 — APPROVED** [review]
Broadly LGTM. Just fix the minor points here, and let's give this a spin.

File: source/agora/network/RPC.d (hunk 5 above)
Comment (omerfirmak) [reply source/agora/network/RPC.d]: Let's just throw on that. If 1024 is enough for our arguments there wont be any problem.

Comment (Geod24) [reply source/agora/network/RPC.d]: Sounds good.

**Review by Geod24 — APPROVED** [review]
LGTM

**Comment by omerfirmak** [comment]
https://github.com/bosagora/agora/pull/2520/checks?check_run_id=3731322944#step:11:8057

@hewison-chris

**Comment by omerfirmak** [comment]
Self merging to ready up a dependant PR

**Pull request merged**